File: sdks/python/tikee/src/tikee/sandbox_tools.py
"""Sandbox tool resolver and installer for Python SDK script runners."""

from __future__ import annotations

import os
import platform
import shutil
import subprocess
from dataclasses import dataclass
from pathlib import Path


def _env_or(key: str, fallback: str) -> str:
    return os.environ.get(key, fallback) or fallback


@dataclass(slots=True)
class SandboxToolResolver:
    """Resolve and optionally install lightweight sandbox tools."""

    state_dir: str = ""
    auto_install: bool = True
    install_timeout: float = 120.0

    def resolve_srt(self) -> tuple[str, bool]:
        return self._resolve_tool("srt", "srt", lambda d: self._run_installer(self._managed_bin(d), "npm", "install", "-g", "--prefix", str(d), _env_or("TIKEE_SRT_NPM_PACKAGE", "@anthropic-ai/sandbox-runtime")))

    def resolve_ripgrep(self) -> tuple[str, bool]:
        return self._resolve_tool("rg", "rg", lambda d: self._run_installer(self._managed_bin(d), "cargo", "install", "--root", str(d), "ripgrep"))

    def resolve_deno(self) -> tuple[str, bool]:
        def install(directory: Path) -> None:
            if os.name == "nt":
                self._run_installer(self._managed_bin(directory), "powershell", "-NoProfile", "-ExecutionPolicy", "Bypass", "-Command", "irm https://deno.land/install.ps1 | iex")
            else:
                self._run_installer(self._managed_bin(directory), "sh", "-c", f"curl -fsSL https://deno.land/install.sh | DENO_INSTALL='{directory}' sh")
        return self._resolve_tool("deno", "deno", install)

    def resolve_rhai(self) -> tuple[str, bool]:
        return self._resolve_tool("rhai-run", "rhai-run", lambda d: self._run_installer(self._managed_bin(d), "cargo", "install", "--root", str(d), "rhai", "--bins", "--features", "bin-features"))

    def resolve_powershell(self) -> tuple[str, bool]:
        return self._resolve_tool("pwsh", "pwsh", self._install_powershell)

    def resolve_node(self) -> tuple[str, bool]:
        return self.resolve_interpreter("node")

    def resolve_npm(self) -> tuple[str, bool]:
        return self.resolve_interpreter("npm")

    def resolve_interpreter(self, binary: str) -> tuple[str, bool]:
        path = shutil.which(binary)
        if path and self._command_works(path, "--version"):
            return path, True
        return "", False

    def _resolve_tool(self, key: str, binary: str, installer) -> tuple[str, bool]:
        path = shutil.which(binary)
        if path and self._tool_works(binary, path):
            return path, True
        directory = self._install_dir(key)
        local = self._managed_bin(directory) / self._executable_name(binary)
        if self._tool_works(binary, str(local)):
            return str(local), True
        if not self.auto_install:
            return str(local), False
        try:
            installer(directory)
        except Exception:
            return str(local), False
        return str(local), self._tool_works(binary, str(local))

    def _install_dir(self, key: str) -> Path:
        base = Path(self.state_dir.strip()) if self.state_dir.strip() else Path.home() / ".tikee"
        return base / "sandbox-tools" / key

    @staticmethod
    def _managed_bin(directory: Path) -> Path:
        return directory / "bin"

    @staticmethod
    def _executable_name(binary: str) -> str:
        return f"{binary}.exe" if os.name == "nt" else binary

    def _run_installer(self, managed_bin: Path, *command: str) -> None:
        managed_bin.mkdir(parents=True, exist_ok=True)
        env = os.environ.copy()
        env["PATH"] = os.pathsep.join(_dedupe([str(managed_bin), *env.get("PATH", "").split(os.pathsep)]))
        subprocess.run(command, check=True, timeout=self.install_timeout, env=env)

    def _install_powershell(self, directory: Path) -> None:
        if os.name == "nt":
            self._run_installer(self._managed_bin(directory), "winget", "install", "-e", "--id", "Microsoft.PowerShell")
            return
        machine = platform.machine().lower()
        system = platform.system().lower()
        platform_key = {
            ("linux", "x86_64"): "linux-x64",
            ("linux", "amd64"): "linux-x64",
            ("linux", "aarch64"): "linux-arm64",
            ("darwin", "x86_64"): "osx-x64",
            ("darwin", "arm64"): "osx-arm64",
        }.get((system, machine))
        if not platform_key:
            raise RuntimeError(f"PowerShell auto-install does not support {system}/{machine}")
        version = _env_or("TIKEE_POWERSHELL_VERSION", "7.5.4")
        archive_name = f"powershell-{version}-{platform_key}.tar.gz"
        url = _env_or("TIKEE_POWERSHELL_DOWNLOAD_URL", f"https://github.com/PowerShell/PowerShell/releases/download/v{version}/{archive_name}")
        bin_dir = self._managed_bin(directory)
        extract_dir = directory / f"powershell-{version}"
        bin_dir.mkdir(parents=True, exist_ok=True)
        extract_dir.mkdir(parents=True, exist_ok=True)
        archive = directory / archive_name
        self._run_installer(bin_dir, "curl", "-fsSL", url, "-o", str(archive))
        self._run_installer(bin_dir, "tar", "-xzf", str(archive), "-C", str(extract_dir))
        pwsh = extract_dir / "pwsh"
        pwsh.chmod(0o755)
        link = bin_dir / "pwsh"
        if link.exists() or link.is_symlink():
            link.unlink()
        try:
            link.symlink_to(pwsh)
        except OSError:
            shutil.copy2(pwsh, link)
            link.chmod(0o755)

    def _command_works(self, command: str, *args: str) -> bool:
        if os.sep in command and not Path(command).exists():
            return False
        try:
            subprocess.run([command, *args], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, timeout=2.0, check=True)
            return True
        except Exception:
            return False

    def _tool_works(self, binary: str, command: str) -> bool:
        if binary == "srt":
            return self._command_works(command, "--version") or self._command_works(command, "--help")
        if binary == "rhai-run":
            return self._command_works(command, "--help") or self._command_works(command, "--version")
        return self._command_works(command, "--version")


def _dedupe(values: list[str]) -> list[str]:
    out: list[str] = []
    for value in values:
        if value and value not in out:
            out.append(value)
    return out

File: sdks/nodejs/tikeo/src/script/sandboxTools.ts
import { existsSync, mkdirSync, rmSync, chmodSync, symlinkSync, copyFileSync, writeFileSync, mkdtempSync, renameSync } from "node:fs";
import { homedir, platform, arch, tmpdir } from "node:os";
import { join } from "node:path";
import { spawn, spawnSync } from "node:child_process";

const backgroundInstalls = new Set<string>();

export class SandboxToolResolver {
  constructor(public stateDir = "", public autoInstall = true, public installTimeoutMs = 120_000, public requireManagedTools = false) {}

  resolveSrt(): [string, boolean] {
    return this.resolveTool("srt", "srt", (dir) => this.runInstaller(this.managedBin(dir), "npm", ["install", "-g", "--prefix", dir, process.env.TIKEO_SRT_NPM_PACKAGE || "@anthropic-ai/sandbox-runtime"]));
  }
  resolveRipgrep(): [string, boolean] { return this.resolveTool("rg", "rg", (dir) => this.runInstaller(this.managedBin(dir), "cargo", ["install", "--root", dir, "ripgrep"])); }
  resolveDeno(): [string, boolean] {
    return this.resolveTool("deno", "deno", (dir) => {
      if (process.platform === "win32") return this.runInstaller(this.managedBin(dir), "powershell", ["-NoProfile", "-ExecutionPolicy", "Bypass", "-Command", "irm https://deno.land/install.ps1 | iex"]);
      return this.runInstaller(this.managedBin(dir), "sh", ["-c", `curl -fsSL https://deno.land/install.sh | DENO_INSTALL='${dir}' sh`]);
    });
  }
  resolveRhai(): [string, boolean] { return this.resolveTool("rhai-run", "rhai-run", (dir) => this.runInstaller(this.managedBin(dir), "cargo", ["install", "--root", dir, "rhai", "--bins", "--features", "bin-features"])); }
  resolvePowerShell(): [string, boolean] { return this.resolveTool("pwsh", "pwsh", (dir) => this.installPowerShell(dir)); }
  resolveNode(): [string, boolean] { return this.resolveInterpreter("node"); }
  resolveNpm(): [string, boolean] { return this.resolveInterpreter("npm"); }

  resolveInterpreter(binary: string): [string, boolean] {
    if (this.requireManagedTools) {
      const path = join(this.managedBin(this.installDir(binary)), executableName(binary));
      return this.commandWorks(path, ["--version"]) || (binary === "sh" && this.commandWorks(path, ["-c", "exit 0"])) ? [path, true] : ["", false];
    }
    const path = findOnPath(binary);
    return path && this.commandWorks(path, ["--version"]) ? [path, true] : ["", false];
  }

  private resolveTool(key: string, binary: string, installer: (dir: string) => void | Promise<void>): [string, boolean] {
    if (!this.requireManagedTools) {
      const found = findOnPath(binary);
      if (found && this.toolWorks(binary, found)) return [found, true];
    }
    const legacyDir = this.legacyInstallDir(key);
    if (legacyDir) {
      const legacyLocal = join(this.managedBin(legacyDir), executableName(binary));
      if (this.toolWorks(binary, legacyLocal)) return [legacyLocal, true];
    }
    const dir = this.installDir(key);
    const local = join(this.managedBin(dir), executableName(binary));
    if (this.toolWorks(binary, local)) return [local, true];
    if (!this.autoInstall) return [local, false];
    this.scheduleBackgroundInstall(key, binary, dir, installer);
    return [local, false];
  }


  private scheduleBackgroundInstall(key: string, binary: string, dir: string, installer: (dir: string) => void | Promise<void>): void {
    const installKey = `${key}@${dir}`;
    if (backgroundInstalls.has(installKey)) return;
    backgroundInstalls.add(installKey);
    setImmediate(async () => {
      try {
        await installer(dir);
      } catch (error) {
        console.warn(`[tikeo.sandbox] background auto-install failed tool=${binary} error=${error instanceof Error ? error.message : String(error)}`);
        return;
      }
      const local = join(this.managedBin(dir), executableName(binary));
      if (!this.toolWorks(binary, local)) console.warn(`[tikeo.sandbox] background auto-install completed but tool is still unavailable tool=${binary} path=${local}`);
    });
  }

  private installDir(key: string): string { return join(hostSandboxToolsRoot(), key); }
  private legacyInstallDir(key: string): string { return this.stateDir.trim() ? join(this.stateDir.trim(), "sandbox-tools", key) : ""; }
  private managedBin(dir: string): string { return join(dir, "bin"); }

  private runInstaller(managedBin: string, command: string, args: string[]): Promise<void> {
    return this.runInstallerWithTimeout(this.installTimeoutMs, managedBin, command, args);
  }

  private runInstallerWithTimeout(timeoutMs: number, managedBin: string, command: string, args: string[]): Promise<void> {
    mkdirSync(managedBin, { recursive: true });
    const env = { ...process.env, PATH: dedupe([managedBin, ...(process.env.PATH ?? "").split(process.platform === "win32" ? ";" : ":")]).join(process.platform === "win32" ? ";" : ":") };
    return new Promise((resolve, reject) => {
      const child = spawn(command, args, { stdio: "inherit", env });
      const timer = setTimeout(() => {
        child.kill("SIGKILL");
        reject(new Error(`installer timed out: ${command}`));
      }, timeoutMs);
      child.once("error", (error) => {
        clearTimeout(timer);
        reject(error);
      });
      child.once("exit", (code) => {
        clearTimeout(timer);
        if (code === 0) resolve();
        else reject(new Error(`installer failed: ${command} exit=${code ?? "signal"}`));
      });
    });
  }

  private async installPowerShell(dir: string): Promise<void> {
    if (process.platform === "win32") { await this.runInstaller(this.managedBin(dir), "winget", ["install", "-e", "--id", "Microsoft.PowerShell"]); return; }
    const key = powerShellArchivePlatform();
    if (!key) throw new Error(`PowerShell auto-install does not support ${platform()}/${arch()}`);
    const version = process.env.TIKEO_POWERSHELL_VERSION || "7.5.4";
    const name = `powershell-${version}-${key}.tar.gz`;
    const url = process.env.TIKEO_POWERSHELL_DOWNLOAD_URL || `https://github.com/PowerShell/PowerShell/releases/download/v${version}/${name}`;
    const bin = this.managedBin(dir);
    mkdirSync(dir, { recursive: true });
    const releaseLock = acquireInstallLock(dir);
    try {
      const link = join(bin, "pwsh");
      if (this.toolWorks("pwsh", link)) return;
      const tmp = mkdtempSync(join(dir, ".pwsh-install-"));
      try {
        const archive = join(dir, name);
        const partialArchive = join(dir, `${name}.part`);
        const tmpArchive = join(tmp, name);
        const tmpExtract = join(tmp, "extract");
        const finalExtract = join(dir, `powershell-${version}`);
        mkdirSync(bin, { recursive: true }); mkdirSync(tmpExtract, { recursive: true });
        if (existsSync(archive)) {
          copyFileSync(archive, tmpArchive);
        } else {
          await this.runInstallerWithTimeout(powerShellInstallTimeout(this.installTimeoutMs), bin, "curl", ["-fL", "-C", "-", url, "-o", partialArchive]);
          copyFileSync(partialArchive, tmpArchive);
        }
        await this.runInstallerWithTimeout(powerShellInstallTimeout(this.installTimeoutMs), bin, "tar", ["-xzf", tmpArchive, "-C", tmpExtract]);
        const pwsh = join(tmpExtract, "pwsh");
        if (!existsSync(pwsh)) throw new Error("PowerShell archive did not contain pwsh");
        chmodSync(pwsh, 0o755);
        rmSync(finalExtract, { force: true, recursive: true });
        renameSync(tmpExtract, finalExtract);
        const installedPwsh = join(finalExtract, "pwsh");
        rmSync(link, { force: true });
        rmSync(partialArchive, { force: true });
        try { symlinkSync(installedPwsh, link); } catch { copyFileSync(installedPwsh, link); chmodSync(link, 0o755); }
      } finally {
        rmSync(tmp, { force: true, recursive: true });
      }
    } finally {
      releaseLock();
    }
  }

  private commandWorks(command: string, args: string[]): boolean {
    if ((command.includes("/") || command.includes("\\")) && !existsSync(command)) return false;
    const result = spawnSync(command, args, { stdio: "ignore", timeout: 2_000 });
    return result.status === 0;
  }

  private toolWorks(binary: string, command: string): boolean {
    if (binary === "srt") return this.commandWorks(command, ["--version"]) || this.commandWorks(command, ["--help"]);
    if (binary === "rhai-run") return this.rhaiWorks(command);
    return this.commandWorks(command, ["--version"]);
  }

  private rhaiWorks(command: string): boolean {
    if ((command.includes("/") || command.includes("\\")) && !existsSync(command)) return false;
    const dir = mkdtempSync(join(tmpdir(), "tikeo-rhai-probe-"));
    const script = join(dir, "probe.rhai");
    try {
      writeFileSync(script, 'print("tikeo-rhai-probe");\n');
      return this.commandWorks(command, [script]);
    } finally {
      rmSync(dir, { recursive: true, force: true });
    }
  }
}

function hostSandboxToolsRoot(): string {
  return (process.env.TIKEO_SANDBOX_TOOLS_DIR || "").trim() || join(homedir(), ".tikeo", "sandbox-tools");
}
function acquireInstallLock(dir: string): () => void {
  const lockDir = join(dir, ".install.lock");
  const deadline = Date.now() + 120_000;
  while (true) {
    try {
      mkdirSync(lockDir);
      return () => rmSync(lockDir, { force: true, recursive: true });
    } catch {
      if (Date.now() >= deadline) throw new Error(`timed out waiting for sandbox tool install lock: ${lockDir}`);
      Atomics.wait(new Int32Array(new SharedArrayBuffer(4)), 0, 0, 100);
    }
  }
}

function powerShellInstallTimeout(timeoutMs: number): number {
  const configured = (process.env.TIKEO_POWERSHELL_INSTALL_TIMEOUT_MILLIS || "").trim();
  if (configured) {
    const parsed = Number.parseInt(configured, 10);
    if (Number.isFinite(parsed) && parsed > 0) return Math.max(1_000, parsed);
  }
  return Math.max(timeoutMs, 1_800_000);
}

function executableName(binary: string): string { return process.platform === "win32" ? `${binary}.exe` : binary; }
function findOnPath(binary: string): string | undefined {
  const paths = (process.env.PATH ?? "").split(process.platform === "win32" ? ";" : ":");
  for (const p of paths) {
    const full = join(p, executableName(binary));
    if (existsSync(full)) return full;
  }
  return undefined;
}
function powerShellArchivePlatform(): string {
  const key = `${platform()}/${arch()}`;
  return ({ "linux/x64": "linux-x64", "linux/arm64": "linux-arm64", "darwin/x64": "osx-x64", "darwin/arm64": "osx-arm64" } as Record<string, string>)[key] ?? "";
}
function dedupe(values: string[]): string[] { return [...new Set(values.filter(Boolean))]; }
